javascript/ql/src/semmle/javascript/ExtendedStaging.qll

```diff
@@ -0,0 +1,333 @@
+ *
+ * Computing less stages can improve performance as each stages is less likely to recompute non-cached predicates.
+ *
+ * A number of stages are grouped into an extended stage.
+ * An extended stage contains a number of substages - corrosponding to to how the stages would be grouped if this file didn't exist.
+ * Each extended stage is identified by a `cached module` in the `ExtendedStaging` module.
+ *
+ * The number of stages are reduced by using how the compiler groups predicates into stages.
+ * The compiler will group mutually recursive cached predicates, or cached predicates within the same `cached module`, into the same stage.
+ * This file uses the latter by creating a `cached module` with two predicates for each extended stage.
+ * The first predicate is referenced from all the `cached` predicates we want in the same extended stage,
+ * and the second predicate has references to all the `cached` predicates we want in the same extended stage.
+ *
+ * With these two predicates in a `cached module` we ensure that all substages will be in a single stage at runtime.
+ *
+ * Grouping stages into extended stages can cause unnecessary computation, as a concrete query might not depend on
+ * all the substages in an extended stage.
+ * Care should therefore be taken not to group stages into an extended stage, if it is likely that a query only depend
+ * on some but not all the stages in the extended stage.
+ */
+
+import javascript
+private import internal.StmtContainers
+private import semmle.javascript.dataflow.internal.PreCallGraphStep
+private import semmle.javascript.dataflow.internal.FlowSteps
+
+/**
+ * Contains a `cached module` for each extended stage.
+ * Each `cached module` ensures that predicates that are supposed to be in the same stage, are in the same stage.
+ *
+ * Each `cached module` contain two predicates:
+ * The first, `ref`, always holds, and is referenced from `cached` predicates in each of the substages.
+ * The second, `backref`, contains references to `cached` predicate from each substage.
+ * The `backref` predicate starts with `1 = 1 or` to ensure that the predicate will be optimized down to a constant by the optimizer.
+ */
+module ExtendedStaging {
+  /**
+   * The `ast` extended stage.
+   * Consists of 7 substages (as of writing this).
+   *
+   * substage 1:
+   *   AST::ASTNode::getParent
+   * substage 2:
+   *   JSDoc::Documentable::getDocumentation
+   * substage 3:
+   *   StmtContainers::getStmtContainer
+   * substage 4:
+   *   AST::StmtContainer::getEnclosingContainer
+   * substage 5:
+   *   AST::ASTNode::getTopLevel
+   * substage 6:
+   *   AST::isAmbientTopLevel
+   * substage 7:
+   *   Expr::Expr::getStringValue // maybe doesn't belong here?
+   * substage 8:
+   *   AST::ASTNode::isAmbientInternal
+   */
+  cached
+  module Ast {
+    /**
+     * Always holds.
+     * Ensures that a predicate is evaluated as part of the Ast stage.
+     */
+    cached
+    predicate ref() { 1 = 1 }
```

Comment:
No, the behavior would be the same if I used `any()` instead. 
The optimizer compiles it down to a constant in both cases.   

I just preferred `1 = 1`, partially because the syntax highlighting sticks out more. 
(If I used `any()`, then the `any()` would look similar to the other references in the `backref` predicate).